internal/uploader/events_uploader.go

```diff
@@ -0,0 +1,261 @@
+package uploader
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"mime/multipart"
+	"net/http"
+	"net/textproto"
+
+	"github.com/go-openapi/strfmt"
+	"github.com/openshift/assisted-service/internal/common"
+	eventsapi "github.com/openshift/assisted-service/internal/events/api"
+	"github.com/openshift/assisted-service/models"
+	"github.com/openshift/assisted-service/pkg/k8sclient"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"gorm.io/gorm"
+)
+
+var _ Client = &eventsUploader{}
+
+type eventsUploader struct {
+	Config
+	db     *gorm.DB
+	log    log.FieldLogger
+	client k8sclient.K8SClient
+}
+
+func (e *eventsUploader) UploadEvents(ctx context.Context, clusterID *strfmt.UUID, eventsHandler eventsapi.Handler) error {
+	pullSecret, err := getPullSecret(e.db, clusterID, e.client, openshiftTokenKey)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get pull secret to upload event data for cluster %s", *clusterID)
+	}
+	buffer, err := prepareFiles(ctx, e.db, clusterID, eventsHandler, pullSecret)
+	if err != nil {
+		return errors.Wrapf(err, "failed to prepare files to upload for cluster %s", *clusterID)
+	}
+	formBuffer, contentType, err := prepareBody(buffer)
+	if err != nil {
+		return errors.Wrapf(err, "failed to prepare event data content to upload for cluster %s", clusterID)
+	}
+	req, err := http.NewRequest(http.MethodPost, e.DataUploadEndpoint, formBuffer)
```

Comment:
Thanks for the suggestion! This request actually goes to a 3rd party service - I'm not familiar with `NewRequestWithContext`, is there a benefit to using `NewRequestWithContext` even if our service is not on the receiving end of this request?

Reply:
no.

TL;DR: 
See https://pkg.go.dev/net/http#NewRequestWithContext

> For an outgoing client request, the context controls the entire lifetime of a request and its response: obtaining a connection, sending the request, and reading the response headers and body

If the context is canceled, it will terminate the request, causing the entire flow to end.
Since now this one is running on its own go routine, it should probably be decoupled from the main flow and should not be executed with the main flow context.

Reply:
Thanks for the explanation! I'll leave this as it is then?